include/npcomp/Dialect/Torch/Transforms/Passes.td

```diff
@@ -164,4 +164,74 @@ def ReduceOpVariants : Pass<"torch-reduce-op-variants", "FuncOp"> {
   }];
 }
 
+def MaximizeValueSemantics : Pass<"torch-maximize-value-semantics", "FuncOp"> {
+  let summary = "Use value-semantic tensors where possible.";
+  let description = [{
+    Use value-semantic tensors where possible to make the program more
+    analyzable by later passes (also, backends prefer value semantics as well).
+
+    This pass is analogous to an SSA-formation pass in a
+    traditional compiler, with the added complication that arrays can alias
+    each other in interesting ways.
+
+    The current code doesn't implement any fancy algorithm, and is intended
+    to be just sufficient for a first e2e spike. An algorithm inspired by the
+    SSA formation literature will need to be implemented.
+
+    Also, this pass doesn't currently handle interprocedural rewriting
+    (of private functions), which is even more complex.
+  }];
+  let constructor = "mlir::NPCOMP::Torch::createMaximizeValueSemanticsPass()";
+}
+
+
+def RefinePublicReturn : Pass<"torch-refine-public-return", "ModuleOp"> {
+  let summary = "Refine public return";
+  let constructor = "mlir::NPCOMP::Torch::createRefinePublicReturnPass()";
+  let description = [{
+    Refines types of values returned from public functions based on
+    intraprocedural information.
+
+    This pass effectively encodes an assumption by the pass pipeline author that
+    the public calling convention of the module can have its types refined,
+    without causing ABI mismatches. This is frequently true -- for example, in
+    many systems, `tensor<?x?xf32>`, `tensor<3x3xf32>` and
```

Comment:
Did you mean to be using the build-in `tensor` nomenclature here?

Reply:
Good catch!